### ISSUE TITLE
FIeldnames cannot contain [ if value contains & character (e.g. <p>apple &amp; orange</p>)

### ISSUE THREAD

**Issue by hailwood** [silverstripe/silverstripe-framework#5277]
TLDR: field values are truncated at an ampersand if the fieldname contains [

**Expected Output**

``` html
"<p>apple &amp; banana</p>"
```

**Actual Output**

``` html
"<p>apple "
```

**How to reproduce**

add the following to a form on a dataobject.

``` php
HtmlEditorField::create('Geocontent[AU]', 'Content for Australia');
```

In the form set the value to "apple & banana" - html view should show

``` html
<p>apple &amp; banana</p>
```

In onbeforeWrite attempt to read the value

``` php
$val = $this->{"Geocontent[AU]"};
die($val);
```

**What's happening?**
This is the offending line https://github.com/silverstripe/silverstripe-framework/blob/3ee8f505b75af39817d394bcf48623e092f43ed2/forms/Form.php#L1431

From what I can tell it's splitting the string on the `&` to get the parts, but due to `Title=Awesome Page&GeoContent[AU]=<p>Apple &amp; Orange</p>` it's truncating the data.

**Comment by tractorcow** [comment]
Question : Does it matter if the [] has a value? e.g.

`Geocontent[]` vs `Geocontent[AU]`? Perhaps the code works for the former, but not the latter?

**Comment by hailwood** [comment]
Hey @tractorcow I don't believe so as the part that causes the issue occurs based purely on the existence of [ in the name.

My thoughts are the intention was for this feature to be used for checkboxes etc where the value would be 1/0 

I've gotted around the issue by changing my fieldname to just GeoContent_AU but I do much prefer the other syntax mostly for the post value of `$gc = $request->postVar('GeoContent'); echo $gc['AU'];` so would be good to get it fixed, but if not then at a minimum I believe any field that can contain an non-numeric/boolean value should throw a warning/error if the fieldname contains [.

**Comment by quamsta** [comment]
This causes major issues with ManyManyExtraFields added to the getCMSFields FieldList: eg:

`$fields->addFieldToTab('Root.Main', HTMLEditorField::create(
				'ManyMany[ArticleNote]'));`

**Comment by tractorcow** [comment]
I see, well we are probably going to deprecate many_many_extraFields, now that we have many_many through relations.

**Comment by dhensby** [comment]
why are we trying to manually parse form data like this? https://github.com/silverstripe/silverstripe-framework/blob/3ee8f505b75af39817d394bcf48623e092f43ed2/forms/Form.php#L1431

I'd imagine it's not just `&amp;` it's any encoded char that has an `&` in it...

**Comment by dhensby** [comment]
OK - can I ask why you'd be using this array notation if you didn't want the data back as an array?

if you had a field like `TextField::create('test[]')` this would break because the data type would be an array and textfield can't load an array.

If you did `TextField::create('test[key])` then this is slightly better, we can pull out the value and populate the field in this example, but again, this is strange and array notation should be used for a field that can hold an array value (like a `checkboxsetfield`)... So this really looks like a misuse of form field names (wouldn't one struggle to do this in vanilla php?) rather than an all out bug in SS.

If you want to donate subfields, you should probably consider dot notatation... `TextField::create('Test.T')`

**Comment by dhensby** [comment]
Yep, ok so if I create a file called `test.php` and this is the contents:

```php
<?php

$value = '';
if ($_POST) {
	$value = array_key_exists('test[]', $_POST) ? htmlentities($_POST['test[]']) : '';
	var_export($_POST);
}


?>
<form method="post" action="test.php">
	<input name="test[]" value="<?=$value?>" />
	<button type="submit">Submit</button>
</form>
```

This is the output:

```html
array (
  'test' => 
  array (
    0 => 'test',
  ),
)<form method="post" action="test.php">
	<input name="test[]" value="" />
	<button type="submit">Submit</button>
</form>
```

I also tested using `test[t]` as the field name and results are the same. 

So you can't name fields like that in PHP and expect to access them with the same name because PHP treats data submitted with `[]` in the name as special array data so you have to have a field called `test` that can handle array values.

I don't think SS should even be trying to access the nested data in the way the code suggests...

**Comment by dhensby** [comment]
I've tried to reproduce this error on 3.4.x-dev and I can't get anything to load into the dataobject as `$this->{'arraylike[field]'}`.

I generally think this approach is wrong and should be discarded as developer error, however I've worked on a patch that could solve the issue and is #6375

I've tested it with xdebug and it now loads values as expected into the `FormField`s, but (as above) I can't get them loaded into the dataobject.

**Comment by quamsta** [comment]
Thanks for your work on this @dhensby , we have a newer site that uses a few manyMany_extraFields with a lot of data that we'd find it hard to migrate to a new many-to-many through relationship at this stage.

**Comment by dhensby** [comment]
fixed by #6375
